CI: cross platform re-implementation as github actions

arummler opened the pull request:
Summary of changes:
- CI implemented for linux gcc and clang, macos clang and windows msvc using github actions
- Improved C++ version detection using ROOT C++ version
- Fixed FileSystem linking
- fixed test on Windows
- removal of appveyor

simonspa commented on CMakeLists.txt:
```diff
@@ -2,6 +2,7 @@ CMAKE_MINIMUM_REQUIRED(VERSION 3.6.3 FATAL_ERROR)
 IF(COMMAND CMAKE_POLICY)
     CMAKE_POLICY(SET CMP0003 NEW) # change linker path search behaviour
     CMAKE_POLICY(SET CMP0048 NEW) # set project version
+    CMAKE_POLICY(SET CMP0057 NEW) # enable IN_LIST
```
Please observe that we had to remove all linker-related Camps in https://github.com/eudaq/eudaq/pull/736 due to issues with ROOT dictionary generation of the StdEvtMonitor.

arummler replied:
CMP0057 is not linker related. It only activated evaluation of the IN_LIST operator.

arummler replied:
I need in Platform.cmake. In principal one get also formulate the check differently but it gets quite ugly.

arummler commented:
@simonspa I believe we could merge this one to get working CI back also on Mac and Windows. Clearly fixes of some Windows non-compiling stuff (mostly from the time when ROOT 6 was not yet available on Window) has to be fixed later. Also trying to add some python wheel with cibuildwheel later.

arummler commented:
Appveyor failing as I did not fix it but replaced completely with github actions. It is removed as part of this MR but still needs to run here.

arummler commented:
Can I merge it? I would like to re-run the CI before addressing the other MRs.

simonspa commented on gui/CMakeLists.txt:
```diff
@@ -31,7 +31,7 @@ qt5_wrap_ui(euRun_UISrcs ui/euRun.ui)
 list(APPEND RUN_GUI_SRC ${euRun_UISrcs} ${euRun_MOCSrcs})
 add_executable(euRun ${RUN_GUI_SRC})
-target_link_libraries(euRun ${EUDAQ_CORE_LIBRARY} ${EXTERNAL_LIB_QT})
+target_link_libraries(euRun ${EUDAQ_CORE_LIBRARY} ${EXTERNAL_LIB_QT} ${EUDAQ_CORE_LIBRARY})
```
linking same lib twice

arummler replied:
That was deliberate as there was an issue on Windows. The fixes I will move into a separate MR...it will likely re-appear.

arummler replied:
But removed for the moment.

simonspa reviewed (commented):
There are many changes only tangential to the CI, it would be much easier if it were possible to review them separately.

simonspa commented on user/tlu/cmake/FindZestSC1.cmake:
```diff
@@ -42,67 +46,81 @@ ELSE(WIN32)
-      MESSAGE(STATUS "Could not find ZestSC1 driver package required by tlu producer. Please refer to the documentation on how to obtain the software. Converter will be built anyway.")
-      return()
+#  IF (EXISTS "/afs/desy.de/group/telescopes/tlu/ZestSC1")
+#    MESSAGE(STATUS "Could not find ZestSC1 driver package required by tlu producer; downloading it now via AFS to ./extern ....")
```
Remove commented code

arummler replied:
ok

arummler commented:
Point taken, I had considered to actually also squash but that is clearly not straightforward, the CI debugging required a lot of push cycles. I can try to separate out some of the "just updated versions". But some of the not obviously CI connected changes are because stuff fails the CI and these are the fixes. I can also separate those out into a separate MR but then the CI will be - correctly - partially failing in this MR. I will attempt later or on the WE.

arummler commented:
I will ping you when it is ready.

arummler commented:
Fix for tbcsDESY  in Mac removed, anyhow currently not enabled in CI as there is another issue. Either to be fixed in another MR or to be marked Mac incompatible in cmake.

arummler commented:
Reverted fix of AdeniumConverter on Mac. As there is no way to deactivate it currently the Mac CI is failing. Either to be re-introduced in another MR or to be deactivated permanently as Mac incompatible in cmake.

arummler commented:
cstdint header removed: AIDA TLU not compiling on Windows but currently failing anyhow and no ipbus available. To be re-introduced in separate MR.

arummler commented:
Reverted ROOT linking modernization for ttree CORE lib to be re-introduced in yet another MR.

arummler commented:
@simonspa Now it contains beside new github workflows and removal of appveyor:
-  Improved C++ version detection using ROOT C++ version
-  Fixed FileSystem linking
-  Download TLU1 dependencies from github instead of DESY afs
-  Fixed test on Windows (libraries not found if not installed)

If it is still too massive I can still remove the last three points. Let me know whether you want to squash or whether we leave it as it is. Now I will have to open about 6 MRs which are partially depending on each other.

simonspa commented on user/tlu/cmake/FindTLUFirmware.cmake:
```diff
@@ -1,42 +1,49 @@
+#  ENDIF()
+  FetchContent_Declare(
+    tlu-dependencies
+    GIT_REPOSITORY    https://github.com/eudaq/tlu-dependencies.git
```
Nobody seems to have cared about this, but the reason for placing it on AFS rather than uploading to GitHub is that these are commercial binary drivers that are not supposed to be redistributed by third parties IIRC.

arummler replied:
Anything else that you want to discuss in a separate MR?

simonspa commented:
Looks alright to me.

@arummler please observe that I will not be able to take on a maintaining role in EUDAQ2 in the future as we have moved on to other shores (i.e. Constellation, see https://constellation.pages.desy.de/ and https://indico.cern.ch/event/1439336/contributions/6242542/).

arummler commented:
@simonspa Thanks for the review.  I am aware of your development work but I am not up to date on the status nor when it will be production ready. The base line for my telescopes here at CERN is for the moment eudaq2 and the test beam season will start here mid-April. I have also introduced new users to eudaq2 just this year and also in itk pixel we have only managed to switch  in 2024. So not sure on which time scale another switch will be possible. From past experience likely not in 2025... Just want to say that for my purposes eudaq2 still needs to be maintained for a while. We can discuss that offline how to deal with reviews as we have MRs pending and I am going to submit a few more (and also the hardware part). Worst case we might need to relax the review requirements as we did for some years for eudaq1.
